wasmtime-wasi/src/host_impls.rs

```diff
@@ -92,6 +92,33 @@ pub fn wasmtime_ssp_proc_exit(rval: wasm32::__wasi_exitcode_t) {
     ::std::process::exit(rval as i32)
 }
 
+pub unsafe fn wasmtime_ssp_fd_prestat_get(
+    prestats: *mut host::fd_prestats,
+    fd: host::__wasi_fd_t,
+    buf: *mut host::__wasi_prestat_t,
+) -> host::__wasi_errno_t {
+    host::rwlock_rdlock(&mut (*prestats).lock as *mut host::rwlock);
```

Comment:
This should ultimately be handled with `::std::sync::RwLock` but I was really apprehensive of changing `fd_prestats` struct which ultimately would result in having to change loads more in `wasmtime-wasi` than just a few calls in `syscalls.rs`. I'm really keen to hear your opinion on it though. Perhaps you can find a nicer solution to this problem.

Reply:
I think in the long term we'll want to switch to Rust's `RwLock`, but we should probably wait on that until all the code that uses read-write locks is in Rust, so we can do it all at once.